src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,174 @@
-        return null;
+        checkIndexIsValid(index, size - 1, "Trying to set at nonexistent or negative index");
+        Node<T> nodeAtIndex = findNodeByIndex(index);
+        T oldValue = nodeAtIndex.value;
+        nodeAtIndex.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexIsValid(index, size - 1, "Trying to remove at nonexistent or negative index");
+        if (index == 0) {
+            return removeFirstElement();
+        }
+        if (index == size - 1) {
+            return removeLastElement();
+        }
+        return removeFromMiddle(findNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> temp = head;
+        if (head.value == object || head.value.equals(object)) {
+            removeFirstElement();
+            return true;
+        }
+
+        while (temp != null) {
+            if (temp.value == null || temp.value.equals(object)) {
+                removeFromMiddle(temp);
+                return true;
+            }
+            temp = temp.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> previous;
+        private Node<T> next;
+
+        public Node(Node<T> previous, T value, Node<T> next) {
+            this.previous = previous;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private void checkIndexIsValid(int index, int size, String message) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(message);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        int count = 0;
+        Node<T> searchResult = head;
+        while (count < index) {
+            searchResult = searchResult.next;
+            count++;
+        }
+        return searchResult;
+    }
+
+    private T removeFromMiddle(Node<T> nodeAtIndex) {
+        Node<T> previous = nodeAtIndex.previous;
+        Node<T> next = nodeAtIndex.next;
+        previous.next = next;
+        next.previous = previous;
+        size--;
+        return nodeAtIndex.value;
+    }
+
+    private T removeLastElement() {
+        Node<T> previousToTail = tail.previous;
+        previousToTail.next = null;
+        T oldValue = tail.value;
+        tail = previousToTail;
+        size--;
+        return oldValue;
+    }
+
+    private T removeFirstElement() {
+        Node<T> nextToHead = head.next;
+        T oldValue = head.value;
+        if (nextToHead != null) {
+            nextToHead.previous = null;
+            head = nextToHead;
+        } else {
+            head = null;
+        }
+        size--;
+        return oldValue;
+    }
```

Comment:
I would recommend to create a single `unlink()` method instead of these 3 methods

think about how we can do this - the method should work if we remove tail, or head, or the element from the middle